### PR TITLE
Move derivative parsed stuff out of phase field

### LINKED ISSUE #5039: Move (Derivative)ParsedMaterial into framework
There is nothing that makes these materials specific to the phase_field module. `ParsedMaterial` can be used in many applications to generate material properties that are functions of non-linear variables.  The autodiffed `DeriviativeParsedMaterial` does not necessarily have to represent a free energy either. @tonkmr 

The Doxygen pages for the classes in question are 
- [ParsedMaterial](http://mooseframework.org/docs/doxygen/modules/classParsedMaterial.html)
- [DerivativeParsedMaterial](http://mooseframework.org/docs/doxygen/modules/classDerivativeParsedMaterial.html)

Note that there are a few parent classes that would have to be moved as well (`DerivativeMaterialInterface` and `DerivativeMaterialPropertyNameInterface` are already located in framework).

### PR DESCRIPTION
**Pull request by dschwen**
Moving the derivative parsed material stuff into framework. This is not specific to phase field. Just putting it out here...

Closes #5039

### CODE REVIEW

**Comment by permcody** [comment]
The same test failed on three different test boxes with threads. It looks like you've introduced a threading only type bug?

**Comment by dschwen** [comment]
Yikes, I moved stuff around but I also had to regold two tests that were using a phase field IC. I'll investigate.

**Comment by dschwen** [comment]
Oh dear, it may be worse. Modules doesn't get testes threaded, does it?

**Comment by permcody** [comment]
> Oh dear, it may be worse. Modules doesn't get tested threaded, does it?

Nope, You definitely need `n_threads` copies of material objects in MOOSE. If this isn't happening with this system, it'll need to be fixed!

**Comment by dschwen** [comment]
That is not the issue.

**Comment by dschwen** [comment]
Alrighty. I think I caught the bug. The issue was that I only let thread 0 do the automatic differentiation (and chain rule material property derivatives) and have all other threads just copy the results from thread 0. One pitfall I missed is a parent class pointer in the FunctionMaterialDescriptors (class that manages material property chain rule derivatives). I need to request material property values from MOOSE through the owning "parent" material. Unfortunately I copied the parent pointer to the thread 0 material class to all other threads, which broke the dependency resolution in MOOSE. Resetting the parent pointer to the thread appropriate object fixies the test for me.

**Comment by dschwen** [comment]
I cannot see the private test app results. So I'll just go on ignoring that yellow badge! @brianmoose @permcody

**Comment by brianmoose** [comment]
You have been added to the very special group to allow access.
Also, the current failing tests in "Private Apps" are not related to this PR.

File: framework/src/utils/FunctionMaterialPropertyDescriptor.C [hunk 1]
```diff
@@ -51,6 +51,17 @@ FunctionMaterialPropertyDescriptor::FunctionMaterialPropertyDescriptor(
 {
 }
 
+FunctionMaterialPropertyDescriptor::FunctionMaterialPropertyDescriptor(
+    const FunctionMaterialPropertyDescriptor & rhs, MooseObject * parent)
+  : _fparser_name(rhs._fparser_name),
+    _base_name(rhs._base_name),
+    _dependent_vars(rhs._dependent_vars),
+    _derivative_vars(rhs._derivative_vars),
+    _value(NULL),
```
Comment (permcody) [inline framework/src/utils/FunctionMaterialPropertyDescriptor.C]: NULL? What is this, 2003?

File: framework/src/materials/DerivativeParsedMaterialHelper.C [hunk 2]
```diff
@@ -101,14 +101,16 @@ DerivativeParsedMaterialHelper::assembleDerivatives()
     }
 
     // copy coupled material properties
-    for (unsigned int i = 0; i < master->_mat_prop_descriptors.size(); ++i)
+    auto start = _mat_prop_descriptors.size();
+    for (unsigned int i = start; i < master->_mat_prop_descriptors.size(); ++i)
```
Comment (permcody) [inline framework/src/materials/DerivativeParsedMaterialHelper.C]: OK - but kind of scatterbrained. `auto` followed immediately by `unsigned int`!  If you are going to do this, why not `decltype`? or even better use our macro `beginIndex()`!

**Review by permcody — DISMISSED** [review]
This can go in unless anybody on the @idaholab/moose-team wants to put the smack down.

**Comment by lindsayad** [comment]
I like this in the framework. It beats having to compile phase field and its dependencies when all you want to do is make use of derivative materials. Just the name itself is entirely general

File: framework/src/materials/DerivativeParsedMaterialHelper.C (hunk 2 above)
Comment (dschwen) [reply framework/src/materials/DerivativeParsedMaterialHelper.C]: `beginIndex`?! That would get me `0` which is wrong. I'll do the decltype dance.

File: framework/src/utils/FunctionMaterialPropertyDescriptor.C (hunk 1 above)
Comment (dschwen) [reply framework/src/utils/FunctionMaterialPropertyDescriptor.C]: Dude, I've been writing 6502 assembly off work the past few weeks. This all is science fiction to me!

File: framework/src/materials/DerivativeParsedMaterialHelper.C (hunk 2 above)
Comment (permcody) [reply framework/src/materials/DerivativeParsedMaterialHelper.C]: Nope `beginIndex` takes multiple paramters, there's this form too:
`beginIndex(container, start_index)`

Comment (dschwen) [reply framework/src/materials/DerivativeParsedMaterialHelper.C]: Oh, there is a two argument version of that macro...

Comment (permcody) [reply framework/src/materials/DerivativeParsedMaterialHelper.C]: So I'm not insisting you use that method (I personally like it), but most people prefer to just declare the wrong type directly instead (e.g. `unsigned int`).

File: framework/src/utils/FunctionMaterialPropertyDescriptor.C (hunk 1 above)
Comment (dschwen) [reply framework/src/utils/FunctionMaterialPropertyDescriptor.C]: FYI, those files were written in 2015(!). I've just pushed a commit to update them. There might be more to do in the files you didn't comment on...

**Review by permcody — APPROVED** [review]
Well you showed me!

**Comment by dschwen** [comment]
😕

**Pull request merged**